### PR TITLE
WT-7748 cmake: fix library probes for libs not on the default linker path

### PR DESCRIPTION
**Pull request by sauclovian-wt**
On a lot of systems external libraries are in places like /usr/local
or /usr/pkg or /opt that aren't in the default linker path. Use
find_library() instead of check_library_exists() so they can be found
according to user-supplied cmake paths.

When linking with the library, use the path found instead of just the
library name. This is what the cmake manual recommends, rather than
adding to the library path and generating -L linker options.

Also add a HEADER option to config_lib(), use it to look for the
library's header files, and add the resulting dir to the include
path. Pass a header name to uses of config_lib() where this makes
sense.

And, print what's happening in config_lib() since otherwise it's
completely silent.

### CODE REVIEW

**Comment by lukech** [comment]
autobuild: Can one of the admins verify this patch?

File: build_cmake/helpers.cmake
Comment (alisonfel) [inline build_cmake/helpers.cmake:431]: We can remove this dead code :)

File: ext/compressors/lz4/CMakeLists.txt
Comment (alisonfel) [inline ext/compressors/lz4/CMakeLists.txt:62]: Can remove line

File: ext/compressors/snappy/CMakeLists.txt
Comment (alisonfel) [inline ext/compressors/snappy/CMakeLists.txt:61]: Can remove line

File: ext/compressors/zstd/CMakeLists.txt
Comment (alisonfel) [inline ext/compressors/zstd/CMakeLists.txt:61]: Can remove line

File: build_cmake/helpers.cmake
Comment (alisonfel) [inline build_cmake/helpers.cmake:429]: I quite like this change! I think its a much better solution than using the `check_library_exists` macro. 

Though, given we are removing this line we can also probably clean up usage of  `${CONFIG_LIB_FUNC}` (as it was only required by  `check_library_exists`). Would it be possible to remove the `if ("${CONFIG_LIB_FUNC}" STREQUAL "")` check just above this (lines 415-418)?

File: build_cmake/configs/auto.cmake
Comment (alisonfel) [inline build_cmake/configs/auto.cmake:277]: On my comment about `CONFIG_LIB_FUNC`, we can remove this line + in the other `config_lib` calls

**Review by alisonfel — COMMENTED** [review]
Thanks so much for the PR, I'm super excited to see some external CMake usage! The changes look pretty good and I'm happy to incorporate the improvements. 
I just have some minor comments on cleaning up bits of dead code.

File: build_cmake/helpers.cmake
Comment (sauclovian-wt) [reply build_cmake/helpers.cmake:431]: I left that there on purpose for reference (along with the function names) in case it's wanted again. There are cases where a function might be in any of several base system libraries, or in libc and available by default, and in those cases you might want a version of config_lib() that does actually use the function name. (I was thinking that in the long run maybe there should be two versions of config_lib(), one that takes a function and one that doesn't, or something like that.) However, currently WT doesn't have any such cases where the function name matters, and maybe if any appear they should be handled by config_func() anyhow. I am not an expert on cmake, so if you think all of this should just go I'm happy to defer to your opinion; just wanted to give my reasoning first.

File: ext/compressors/lz4/CMakeLists.txt
Comment (sauclovian-wt) [reply ext/compressors/lz4/CMakeLists.txt:62]: Indeed, I'll push another commit that gets rid of these. I also just noticed that I forgot to make this change for the zlib extension (on my machine zlib is in the default linker path, so it didn't fail, but that isn't always true), so I'll push that as well.

File: build_cmake/helpers.cmake
Comment (sauclovian-wt) [reply build_cmake/helpers.cmake:429]: See other reply above, but yes, if you think it won't be wanted again in the future I can get rid of CONFIG_LIB_FUNC and the FUNC arguments to config_lib.

**Comment by sauclovian-wt** [comment]
I don't appear to be allowed to see the test failure results, requires a mongodb.com address :-(

**Comment by sauclovian-wt** [comment]
hmm apparently the extra changes made the failures go away? odd...

File: build_cmake/helpers.cmake
Comment (alisonfel) [reply build_cmake/helpers.cmake:431]: Thanks for the background and reasoning on the change! That's completely fair & thanks for the consideration.

On second thought, to save the hassle, I'm cool to leave this code in just for know. I'll be adding some small cleanups in another upcoming PR that will probably involve using the `config_lib`  (with regards to the tcmalloc lib). If its still unnecessary then I will look into cleaning up the function :)

**Comment by alisonfel** [comment]
> hmm apparently the extra changes made the failures go away? odd...

@sauclovian-wt I took a look at the previous failing build and failures were unrelated to your CMake changes (a flaky test failure). So no failures you need to worry about with the CMake changes.

File: build_cmake/helpers.cmake
Comment (sauclovian-wt) [reply build_cmake/helpers.cmake:431]: If at that point you want me to do it, feel free to ping me...

**Comment by alisonfel** [comment]
@sauclovian-wt So sorry for the hold up. I recently discovered a slight issue with these changes when compiling on MacOS (I wanted to do a quick test on my local machine before merging). 

As we now make it easier to link libraries outside the default linker path, Clang can complain with the diagnostic `-Wpoison-system-directories` when it see non-standard include paths e.g `/usr/local/include`. This diagnostic is really only useful for cross compilation (though for some confusing reason the MacOS clang compiler has this warning show up even when not cross-compiling). Note, this only happens when compiling with strict and using the MacOS vendored Clang compiler. Your changes are still great! This is more a general bug in the CMake build system that this PR has made more prevalent. 

I have a really small patch that fixes this, would you mind if I push it as a commit to this PR branch before merging?

**Comment by sauclovian-wt** [comment]
Absolutely go ahead! No idea how to fix that and don't currently have a MacOS dev environment to try it in either...

**Comment by sauclovian-wt** [comment]
LGTM :-)

**Review by alisonfel — APPROVED** [review]
LGTM! Thanks again for the fixes :) 
I'll go ahead and merge this.

**Pull request merged**
